### PR TITLE
adding complete forward and backward pass for ferminet pretraining

### PR DESCRIPTION
**Pull request by shaipranesh2**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/ferminet.py [hunk 1]
```diff
@@ -71,6 +65,8 @@ def __init__(self,
 
         Attributes
         ----------
+        running_diff: torch.Tensor
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:68]: Adding an attribute tensor to keep track of running loss

File: deepchem/models/torch_models/ferminet.py [hunk 2]
```diff
@@ -148,6 +145,30 @@ def forward(self, input) -> torch.Tensor:
             0].forward(one_electron, one_electron_vector_permuted)
         return psi
 
+    def loss(self,
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:148]: loss functions, right now its implemented for pertaining (MSELoss)

File: deepchem/models/torch_models/ferminet.py [hunk 3]
```diff
@@ -205,10 +226,8 @@ def __init__(self,
         Torch tensor containing electrons for each atom in the nucleus
     molecule: ElectronSampler
         ElectronSampler object which performs MCMC and samples electrons
-    loss_value: Optional[torch.Tensor] (default None)
+    loss_value: torch.Tensor (default torch.tensor(0))
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:229]: attribute that keeps track of the last loss value

File: deepchem/models/torch_models/ferminet.py [hunk 4]
```diff
@@ -274,13 +292,14 @@ def __init__(self,
...
                               nuclear_charge=torch.Tensor(charge),
-                              batch_size=self.batch_no)
+                              batch_size=self.batch_no).float()
 
         self.molecule: ElectronSampler = ElectronSampler(
             batch_no=self.batch_no,
             central_value=self.nucleon_pos,
             seed=self.seed,
-            f=lambda x: test_f(x),  # Will be replaced in successive PR
+            f=lambda x: self.random_walk(x
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:301]: changing the sampling function to the random walk function which returns the log wavefucntion probability

File: deepchem/models/torch_models/ferminet.py [hunk 5]
```diff
@@ -291,6 +310,34 @@ def __init__(self,
             self.model,
             loss=torch.nn.MSELoss())  # will update the loss in successive PR
 
+    def evaluate_hf(self, x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:313]: Function to evaluate hf orbitals at the given electron coordinates

File: deepchem/models/torch_models/ferminet.py [hunk 6]
```diff
@@ -313,3 +360,59 @@ def prepare_hf_solution(self):
...
+            contains the learning rate for the model fitting
+        weight_decay: float (default: 0.0001)
+            contains the weight_decay for the model fitting
+        burn_in: int (default 100)
+            contains the number of steps for to perform burn-in to get better starting point
+        """
+        optimizer = torch.optim.Adam(self.model.parameters(),
+                                     lr=lr,
+                                     weight_decay=weight_decay)
+        for _ in range(burn_in):
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:408]: Here "burn-in" of a number of samples is taking place (initializing electrons a number of times with the last value being not used), as repeatedly sampling can increase the probability of choosing the most probable electron coordinates to start with.

File: deepchem/models/torch_models/ferminet.py [hunk 7]
```diff
@@ -274,13 +292,14 @@ def __init__(self,
         self.model = Ferminet(nucl,
                               spin=(self.up_spin, self.down_spin),
                               nuclear_charge=torch.Tensor(charge),
-                              batch_size=self.batch_no)
+                              batch_size=self.batch_no).float()
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py]: makes the model weights and biases to be float in

File: deepchem/models/torch_models/ferminet.py [hunk 8]
```diff
@@ -313,3 +360,59 @@ def prepare_hf_solution(self):
         self.mol.build(parse_arg=False)
         self.mf = pyscf.scf.UHF(self.mol)
         _ = self.mf.kernel()
+
+    def random_walk(self, x: np.ndarray) -> np.ndarray:
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:364]: This function gets called at every step of electron sampling, and it returns the probability of the sampled electrons - log of the wavefunction from the model and the HF solution (this is only for pretraining).

File: deepchem/models/torch_models/ferminet.py [hunk 9]
```diff
@@ -313,3 +360,59 @@ def prepare_hf_solution(self):
...
+        output = self.model.forward(x)
+        np_output = output.detach().cpu().numpy()
+        up_spin_mo, down_spin_mo = self.evaluate_hf(x)
+        hf_product = np.prod(
+            np.diagonal(up_spin_mo, axis1=1, axis2=2)**2, axis=1) * np.prod(
+                np.diagonal(down_spin_mo, axis1=1, axis2=2)**2, axis=1)
+        self.model.loss(up_spin_mo, down_spin_mo, pretrain=True)
+        return np.log(hf_product + np_output**2) + np.log(0.5)
+
+    def pretrain(self,
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/torch_models/ferminet.py:386]: We shouldn't have a custom function for pretraining. Instead follow the convention used elsewhere of having a 'task` parameter and add pretraining/supervised tasks for regular training.

**Review by rbharath — COMMENTED** [review]
@shaipranesh2 We shouldn't have a custom pretrain function but rather should have a task field that we can set. See InfoGraph or other models for examples.

We should also expect to see 22 tests passing but I see 21. Are there any new failures?

File: deepchem/models/torch_models/ferminet.py [hunk 10]
```diff
@@ -205,10 +226,12 @@ def __init__(self,
         Torch tensor containing electrons for each atom in the nucleus
     molecule: ElectronSampler
         ElectronSampler object which performs MCMC and samples electrons
-    loss_value: Optional[torch.Tensor] (default None)
+    loss_value: torch.Tensor (default torch.tensor(0))
         torch tensor storing the loss value from the last iteration
-    pretraining_loss_list: List (default [])
-        list with losses for every epoch
+    running_diff: torch.Tensor
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:231]: running diff keeps track of the running average of losses for every sampling move in an epoch. resets to 0 after every epoch

File: deepchem/models/torch_models/ferminet.py [hunk 11]
```diff
@@ -313,3 +368,195 @@ def prepare_hf_solution(self):
...
+                self._optimizer_for_vars[var_key] = (optimizer, lr_schedule)
+        time1 = time.time()
+
+        # Execute the loss function, accumulating the gradients.
+        for _ in range(burn_in):
+            self.molecule.gauss_initialize_position(
+                self.electron_no)  # initialize the position of the electrons
+
+        for _ in range(nb_epoch):
+            if self.task == 'pretraining':
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:502]: Implemented for pretraining now, will exted to 'training' task too

File: deepchem/models/torch_models/ferminet.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py]: done

**Comment by shaipranesh2** [comment]
> @shaipranesh2 We shouldn't have a custom pretrain function but rather should have a task field that we can set. See InfoGraph or other models for examples.
> 
> We should also expect to see 22 tests passing but I see 21. Are there any new failures?

One of the tensorflow CIs throwed an error while installing dependencies, and other than that torch tests are failing due dmpnn. All the ferminet's tests are in DQC and it runs correctly

File: deepchem/models/torch_models/ferminet.py [hunk 12]
```diff
@@ -280,7 +311,8 @@ def __init__(self,
             batch_no=self.batch_no,
             central_value=self.nucleon_pos,
             seed=self.seed,
-            f=lambda x: test_f(x),  # Will be replaced in successive PR
+            f=lambda x: self.random_walk(x
+                                        ),  # Will be replaced in successive PR
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:315]: Make sure to remove the comment in your next update

**Review by rbharath — APPROVED** [review]
LGTM
